### ISSUE TITLE
Deleted/modified messages do not affect emoji or message counts

### ISSUE THREAD

**Issue by flukiluke** [olliequ/KittyBot#23]
It would be nice if they did.

**Comment by dakshAg** [comment]
>  note
        Due to Discord limitations, most message information is unavailable
        during deletion events.


Ref: https://github.com/hikari-py/hikari/blob/master/hikari/events/message_events.py#L627
So we cannot know what emojis were used in the deleted messages unless we keep a record of all messages sent.

**Comment by olliequ** [comment]
One could argue the sending of an emoji in a message is a non-reversible event (so the count should _not_ be decremented). I guess this is to point out that there indeed existed a point in time where the user wanted to use that emoji.

edit: This doesn't cover events where a user accidentally selects the wrong emoji and sends it. Not sure how common this is, though. 

edit2: @dakshAg is it possible in your comment to link the source that contains the text you wrote in-line?

**Comment by olliequ** [comment]
Further thought on this: Hikari is able to track message deletes with: `hikari.MessageDeleteEvent`. I also think the message content can still be seen using: https://www.hikari-py.dev/hikari/events/message_events.html#hikari.events.message_events.MessageDeleteEvent.old_message

Notice how it says there:
> Will be `None` if the message was not found in the cache.

Basically, it's possible if the message is cached, which I think requires Message Intents. No idea if Kitti currently has that. In regards to setting up the 'intents', a Hikari dev told me this:

> You need to do that by providing `intents` kwarg with an hikari.Intents object in hikari.GatewayBot class, as well as setting it in the dev portal. https://hikari-lightbulb.readthedocs.io/en/development/hikari_basics/intents.html

So, a delete event can **definitely** be identified, meaning that the `message_counts` table can be updated. And, it seems the `emoji_counts` table can *also* be updated based on what I just said above with the caching possibilities (as it means the deleted message can be analysed for emojis it contains).

Although, due to the reasons I wrote in the comment above the implementation of this should be thought through to prevent nefarious usage.

**Comment by olliequ** [comment]
On further thought I think this would be nice. It would allow users to undo any damage done by any emoji spamming they do. And this also ties in with #29; counting deleted messages per user for a fun ranking.

**Comment by olliequ** [comment]
Closing this issue as:

If you delete a really old message, its contents (author & message itself) **won't** be able to be retrieved. I think it's related to bot caching. This means the emoji decrement won't work (well, it would, just not on deleting **old** messages), as it means you can't analyse an old messages' deleted messages contents, and so cannot extract the emojis. It would work for the implementation timestamp and onwards.

**Comment by olliequ** [comment]
Reopening upon discovering you *can* get the contents of really old messages using `fetch_message`. It would look something like this:
```
msg = await bot.rest.fetch_message(channel_id, message_id)
bot.cache.set_message(message_id, msg)
```
The flow would be:
**1)** Get ID of just-deleted message (within `GuildMessageDeleteEvent`).
**2)** Fetch message contents using the ID with `fetch_message`.
**3)** Analyse message for emojis. If any, subtract emojis from the users emoji count. 

I don't think the message count should be decremented for 2 reasons:
- Sending a message imo is an irreversible event that should be counted (and definitely made indistinguishable from not sending a message at all). Conversely, emojis should be able to decremented as people spam them to solely boost their count (and there's no reward for sending and removing emojis -- see below)
- People could abuse the party bot thing (e.g. when they hit 10,000 go back to 9,999, hit 10,000 again etc.)

So yes, it seems this issue can be resolved :)

Method: https://www.hikari-py.dev/hikari/api/rest.html#hikari.api.rest.RESTClient.fetch_message

**Comment by olliequ** [comment]
Update: The `fetch_message` will always fail, as you can't fetch something that doesn't exist (and the message doesn't exist as it just got deleted!). Thus to cache really old messages, you could retroactively cache them, but that's gonna be really intensive (constantly tether on the rate limit) as we'd have to retroactively cache over 250,000 messages!

So, this feature can work, but the emoji analysis and decrement would only work on messages that get deleted *after* this feature is implemented (so deleting really old emoji spam messages wouldn't have any effect).

**Comment by olliequ** [comment]
Fixed with #47 -- yay!
